### PR TITLE
Migrate tests to use tokio/collect harness

### PR DESCRIPTION
**Pull request by ddowl**
These commits change all of the test examples in the runtime to use a test harness called `run_link` (name feedback welcome), which encapsulates common behavior in tests of 

1. Setting up collector tasks to drive traffic through the link under test, and
2. Running link's runnables and collector tasks through Tokio.

Consequently, test examples are now more simple/concise due to the lack of boilerplate. Notable improvements can be seen in `CloneLink` and `ClassifyLink` examples.

This is one of the first of many cool possibilities now that we've created a standard `Link` interface, so thanks to @CollinValley and @scgruber for contributing to this effort.

### CODE REVIEW

File: route-rs-runtime/src/composite/m_transform_n_composite.rs [hunk 1]
```diff
@@ -143,57 +143,31 @@ impl<E: Element + Send + 'static> ElementLinkBuilder<E> for MtransformNComposite
...
     fn m_transform_n_composite() {
         let packets = vec![0xDEAD_BEEF, 0xBEEF_DEAD, 0x0A00_0001, 0xFFFF_FFFF];
-        let number_num_egressors = 2;
-        let packet_generator0 = immediate_stream(packets.clone());
-        let packet_generator1 = immediate_stream(packets.clone());
 
         let mut input_streams: Vec<PacketStream<u32>> = Vec::new();
-        input_streams.push(Box::new(packet_generator0));
-        input_streams.push(Box::new(packet_generator1));
+        input_streams.push(immediate_stream(packets.clone()));
```
Comment (CollinValley) [inline route-rs-runtime/src/composite/m_transform_n_composite.rs:157]: This is a nice change, much simpler syntax, focuses the code on what is important for the test

File: route-rs-runtime/src/composite/mton_composite.rs [hunk 2]
```diff
@@ -116,52 +116,29 @@ impl<Packet: Sized + Send + Clone + 'static> LinkBuilder<Packet, Packet> for Mto
...
-        runnables.push(Box::new(collector0));
-
-        run_tokio(runnables);
-
-        let output0: Vec<_> = collector0_output.iter().collect();
-        assert_eq!(output0.len(), packets.len() * 2);
-
-        let output1: Vec<_> = collector1_output.iter().collect();
-        assert_eq!(output1.len(), packets.len() * 2);
+        let results = run_link(link);
```
Comment (CollinValley) [inline route-rs-runtime/src/composite/mton_composite.rs:137]: image
Gotta love it

File: route-rs-runtime/src/link/blackhole_link.rs [hunk 3]
```diff
@@ -122,27 +111,25 @@ mod tests {
...
-            .ingressor(packet_generator0)
-            .ingressor(packet_generator1)
+
+        let link = BlackHoleLink::<IdentityElement<i32>>::new()
+            .ingressor(immediate_stream(packets.clone()))
+            .ingressor(immediate_stream(packets.clone()))
             .build_link();
-        run_tokio(runnables);
+
+        run_link(link);
```
Comment (CollinValley) [inline route-rs-runtime/src/link/blackhole_link.rs:120]: Nit: Check for results.is_empty() here?

File: route-rs-runtime/src/link/blackhole_link.rs [hunk 4]
```diff
@@ -153,43 +140,33 @@ mod tests {
...
-            ExhaustiveCollector::new(0, Box::new(egressors.pop().unwrap()), s0);
-
-        runnables.push(Box::new(link0_port0_collector));
         runnables.append(&mut black_hole_runnables);
 
-        run_tokio(runnables);
-
-        let elem0_port0_output: Vec<_> = link0_port0_collector_output.iter().collect();
-        assert_eq!(elem0_port0_output, vec![0, 2, 420, 4, 6, 8]);
+        let link = (runnables, vec![egressors.pop().unwrap()]);
```
Comment (CollinValley) [inline route-rs-runtime/src/link/blackhole_link.rs:168]: Damn this is pretty slick actually

File: route-rs-runtime/src/link/classify_link.rs [hunk 5]
```diff
@@ -313,112 +298,59 @@ mod tests {
...
-        let even_classifier = ClassifyEvenness::new();
-
         ClassifyLink::new()
             .ingressor(packet_generator)
-            .classifier(even_classifier)
+            .classifier(ClassifyEvenness::new())
             .build_link();
     }
 
+    fn even_classifier(stream: PacketStream<i32>) -> Link<i32> {
```
Comment (CollinValley) [inline route-rs-runtime/src/link/classify_link.rs:307]: Need to build a link? Just return a `Link`! Amazing.

File: route-rs-runtime/src/link/join_link.rs [hunk 6]
```diff
@@ -298,114 +288,63 @@ mod tests {
...
-        let link_output: Vec<_> = collector_output.iter().collect();
-        assert_eq!(link_output.len(), packets.len() * 2);
+        let results = run_link(link);
+        assert_eq!(results[0].len(), packets.len() * 2);
     }
 
+    // TODO: I'm not sure this behavior is appropriate for JoinLink.
+    // Since we know JoinLinks are N to 1 rather than 1 to 1, an `ingressor` function doesn't
+    // really make sense. I read the following example as resetting the single ingressor on a
+    // JoinLink. We should talk about this at the next meetup.
```
Comment (CollinValley) [inline route-rs-runtime/src/link/join_link.rs:319]: Agreed, we just need to decide in the N case, what our policy for the ingressor fn should be.

File: route-rs-runtime/src/utils/test/harness.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+/// "Then" refers to the expected changes to the system due to executing the behavior under test
+/// against the initial context.
+/// This is the state of packet collectors after the input has been exhausted and run through
+/// our Link under test.
+
+/// Since the initial context of "a link's input streams" are coupled to the construction of Links,
+/// let's just expose a function that takes a connected Link, runs it's runnables and collectors
+/// through Tokio, and extracts the output packets into vectors representing egress streams.
+
+// TODO: fn name improvement
```
Comment (CollinValley) [inline route-rs-runtime/src/utils/test/harness.rs:26]: Some other names to consider:
test_link
exercise_link

I don't think run_link is that bad. But if this is a function we only expect to use for tests, then test_link may read more clearly. 
I leave the decision to you, is isn't a big deal

**Review by CollinValley — COMMENTED** [review]
Only one nit
LGTMMD!

"A coder should not be judged by the code they create, rather, the code they remove" --Probably Uncle Bob

**Review by CollinValley — DISMISSED** [review]
Meant to approve!

File: route-rs-runtime/Cargo.toml [hunk 8]
```diff
@@ -9,4 +9,5 @@ license = "MIT"
 tokio = "0.1.22"
 futures = "0.1.29"
 crossbeam = "0.7.2"
-smoltcp = "0.5.0"
\ No newline at end of file
+smoltcp = "0.5.0"
+rand = "*"
```
Comment (moosingin3space) [inline route-rs-runtime/Cargo.toml]: Please pin the dependent crate version.

**Review by CollinValley — APPROVED** [review]
LGTMMD

**Pull request merged**